src/MongoDB.Driver.Core/Core/Compression/Zstd/ZstandardNativeWrapper.cs

```diff
@@ -0,0 +1,292 @@
+*   limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.Core.NativeLibraryLoader;
+
+namespace MongoDB.Driver.Core.Compression.Zstd
+{
+    internal class ZstandardNativeWrapper : IDisposable
+    {
+        #region static
+        public static int MaxCompressionLevel => Zstandard64NativeMethods.ZSTD_maxCLevel();
+        #endregion
+
+        private readonly int _compressionLevel;
+        private readonly CompressionMode _compressionMode;
+        private readonly NativeBufferInfo _inputNativeBuffer = new NativeBufferInfo();
+        private readonly NativeBufferInfo _outputNativeBuffer = new NativeBufferInfo();
+        private bool _operationInitialized;
+        private readonly ulong _recommendedZstreamInputSize;
+        private readonly ulong _recommendedZstreamOutputSize;
+        private IntPtr _zstreamPointer;
+
+        public ZstandardNativeWrapper(CompressionMode compressionMode, int compressionLevel)
+        {
+            _compressionLevel = compressionLevel;
+            _compressionMode = compressionMode;
+
+            switch (_compressionMode)
+            {
+                case CompressionMode.Compress:
+                    // calculate recommended size for input buffer
+                    _recommendedZstreamInputSize = Zstandard64NativeMethods.ZSTD_CStreamInSize();
+                    // calculate recommended size for output buffer. Guarantee to successfully flush at least one complete compressed block
+                    _recommendedZstreamOutputSize = Zstandard64NativeMethods.ZSTD_CStreamOutSize();
+                    _zstreamPointer = Zstandard64NativeMethods.ZSTD_createCStream(); // create resource
+                    break;
+                case CompressionMode.Decompress:
+                    // calculate recommended size for input buffer
+                    _recommendedZstreamInputSize = Zstandard64NativeMethods.ZSTD_DStreamInSize();
+                    // calculate recommended size for output buffer. Guarantee to successfully flush at least one complete block in all circumstances
+                    _recommendedZstreamOutputSize = Zstandard64NativeMethods.ZSTD_DStreamOutSize();
+                    _zstreamPointer = Zstandard64NativeMethods.ZSTD_createDStream(); // create resource
+                    break;
+            }
+        }
+
+        public int RecommendedInputSize => (int)_recommendedZstreamInputSize;
+        public int RecommendedOutputSize => (int)_recommendedZstreamOutputSize;
+
+        // public methods
+        public void Compress(
+            OperationContext operationContext,
+            int compressedSize,
+            int uncompressedSize,
+            out int compressedBufferPosition,
+            out int uncompressedBufferPosition)
+        {
+            Ensure.IsNotNull(operationContext, nameof(operationContext));
```

Comment:
Check that calling `Compress` is valid?

That is, that the `CompressionMode` is `Compress`.